File: modules/purvis-core-expansion/connectorBridge.ts
/**
 * Purvis Core Expansion — Connector Bridge
 * Routes tool/connector requests to APIs, internal functions, or mock connectors.
 *
 * Hosts can register additional connectors at runtime via `registerConnector`.
 */

import { ConnectorRequest, ConnectorResponse } from "./types";
import { config } from "./config";

export type ConnectorHandler = (
  payload: Record<string, unknown>
) => Promise<unknown> | unknown;

const registry = new Map<string, ConnectorHandler>();

/** Register or override a connector handler. */
export function registerConnector(
  type: string,
  handler: ConnectorHandler
): void {
  registry.set(type.toLowerCase(), handler);
}

export function listConnectors(): string[] {
  return [...registry.keys()];
}

/* ---------- Built-in connectors (mocks / safe utilities) ---------- */

registerConnector("echo", (payload) => ({ echoed: payload }));

registerConnector("math", (payload) => {
  const op = String(payload.op ?? "add");
  const a = Number(payload.a ?? 0);
  const b = Number(payload.b ?? 0);
  switch (op) {
    case "add":
      return { result: a + b };
    case "sub":
      return { result: a - b };
    case "mul":
      return { result: a * b };
    case "div":
      return { result: b === 0 ? null : a / b };
    default:
      throw new Error(`Unknown math op: ${op}`);
  }
});

registerConnector("mock", (payload) => ({
  message: "mock connector ok",
  payload,
  at: new Date().toISOString(),
}));

/**
 * Lightweight web/http connector backed by node:fetch (Node 18+).
 * Method defaults to GET, body is JSON-stringified when present.
 * No filesystem, no shell — just an outbound HTTP call.
 */
const httpHandler: ConnectorHandler = async (payload) => {
  const url = String(payload.url ?? "");
  if (!/^https?:\/\//i.test(url)) {
    throw new Error("connector.http: url must start with http:// or https://");
  }
  const method = String(payload.method ?? "GET").toUpperCase();
  const headers =
    (payload.headers as Record<string, string> | undefined) ?? {};
  const body =
    payload.body === undefined
      ? undefined
      : typeof payload.body === "string"
        ? (payload.body as string)
        : JSON.stringify(payload.body);

  const controller = new AbortController();
  const timeout = Number(payload.timeoutMs ?? 5000);
  const t = setTimeout(() => controller.abort(), timeout);
  try {
    const res = await fetch(url, {
      method,
      headers,
      body: body as BodyInit | undefined,
      signal: controller.signal,
    });
    const ct = res.headers.get("content-type") ?? "";
    const data = ct.includes("application/json")
      ? await res.json().catch(() => null)
      : await res.text();
    return { status: res.status, ok: res.ok, data };
  } finally {
    clearTimeout(t);
  }
};
registerConnector("web", httpHandler);
registerConnector("http", httpHandler);

/* ---------- Public entry point ---------- */

export async function executeConnector(
  req: ConnectorRequest
): Promise<ConnectorResponse> {
  if (!config.enableConnectors) {
    return { ok: false, type: req.type, data: null, error: "connectors disabled" };
  }
  const key = String(req.type ?? "").toLowerCase();
  const handler = registry.get(key);
  if (!handler) {
    return {
      ok: false,
      type: req.type,
      data: null,
      error: `unknown connector: ${req.type}`,
    };
  }
  try {
    const data = await handler(req.payload ?? {});
    return { ok: true, type: req.type, data };
  } catch (err) {
    return {
      ok: false,
      type: req.type,
      data: null,
      error: err instanceof Error ? err.message : String(err),
    };
  }
}

File: modules/purvis_core_expansion/connector_bridge.py
"""
Purvis Core Expansion — Connector Bridge
Routes tool/connector requests to APIs, internal functions, or mock connectors.
Hosts can register additional connectors at runtime via register_connector().
"""
from __future__ import annotations
from typing import Callable, Any, Dict, List, Awaitable, Union
import asyncio
import urllib.request
import urllib.error
import json
import datetime as _dt

from .config import config
from .types import ConnectorRequest, ConnectorResponse


ConnectorHandler = Callable[[Dict[str, Any]], Union[Any, Awaitable[Any]]]

_registry: Dict[str, ConnectorHandler] = {}


def register_connector(type_: str, handler: ConnectorHandler) -> None:
    _registry[type_.lower()] = handler


def list_connectors() -> List[str]:
    return list(_registry.keys())


# ---------- Built-in connectors ----------

def _echo(payload: Dict[str, Any]) -> Dict[str, Any]:
    return {"echoed": payload}


def _math(payload: Dict[str, Any]) -> Dict[str, Any]:
    op = str(payload.get("op", "add"))
    a = float(payload.get("a", 0))
    b = float(payload.get("b", 0))
    if op == "add":
        return {"result": a + b}
    if op == "sub":
        return {"result": a - b}
    if op == "mul":
        return {"result": a * b}
    if op == "div":
        return {"result": None if b == 0 else a / b}
    raise ValueError(f"Unknown math op: {op}")


def _mock(payload: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "message": "mock connector ok",
        "payload": payload,
        "at": _dt.datetime.now(_dt.timezone.utc).isoformat(),
    }


def _http(payload: Dict[str, Any]) -> Dict[str, Any]:
    url = str(payload.get("url", ""))
    if not (url.startswith("http://") or url.startswith("https://")):
        raise ValueError("connector.http: url must start with http:// or https://")
    method = str(payload.get("method", "GET")).upper()
    headers = payload.get("headers") or {}
    body = payload.get("body")
    if body is not None and not isinstance(body, (bytes, str)):
        body = json.dumps(body).encode("utf-8")
        headers = {**headers, "Content-Type": "application/json"}
    elif isinstance(body, str):
        body = body.encode("utf-8")
    timeout = float(payload.get("timeoutMs", 5000)) / 1000.0
    req = urllib.request.Request(url, data=body, method=method, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            raw = resp.read()
            ct = resp.headers.get("content-type", "")
            data: Any
            if "application/json" in ct:
                try:
                    data = json.loads(raw.decode("utf-8"))
                except json.JSONDecodeError:
                    data = raw.decode("utf-8", errors="replace")
            else:
                data = raw.decode("utf-8", errors="replace")
            return {"status": resp.status, "ok": 200 <= resp.status < 400, "data": data}
    except urllib.error.HTTPError as e:
        return {"status": e.code, "ok": False, "data": e.reason}


register_connector("echo", _echo)
register_connector("math", _math)
register_connector("mock", _mock)
register_connector("web", _http)
register_connector("http", _http)


# ---------- Public entry ----------

async def execute_connector(req: ConnectorRequest) -> ConnectorResponse:
    if not config.enable_connectors:
        return ConnectorResponse(ok=False, type=req.type, error="connectors disabled")
    handler = _registry.get(req.type.lower())
    if not handler:
        return ConnectorResponse(
            ok=False, type=req.type, error=f"unknown connector: {req.type}"
        )
    try:
        result = handler(req.payload or {})
        if asyncio.iscoroutine(result):
            result = await result
        return ConnectorResponse(ok=True, type=req.type, data=result)
    except Exception as e:  # noqa: BLE001 — surface any handler error
        return ConnectorResponse(ok=False, type=req.type, error=str(e))
